homeassistant/components/zha/core/device.py

```diff
@@ -221,7 +221,10 @@ def is_mains_powered(self) -> bool | None:
     @property
     def device_type(self) -> str:
         """Return the logical device type for the device."""
-        if self._zigpy_device.node_desc is None:
+        if (
+            self._zigpy_device.node_desc is None
+            or self._zigpy_device.node_desc.logical_type is None
```

Comment:
@puddly if there's a node_descriptor, can it have a logical type of None?

Reply:
Note that this can definitely happen in the field. As explained in the linked issue, printing `self._zigpy_device`, `self._zigpy_device.node_desc` and `self._zigpy_device.node_desc.logical_type` outputs respectively in the problematic case:

- `<MijaButton model='lumi.sensor_switch' manuf='LUMI' nwk=0xA327 ieee=00:15:8d:00:04:03:42:17 is_initialized=False>`
- `NodeDescriptor()`
- `None`

Reply:
Also, looking at zigpy source code, it looks like `logical_type` may indeed be `None` (and it is its initial value), *e.g.* in https://github.com/zigpy/zigpy/blob/dev/zigpy/zdo/types.py#L165-L184

Reply:
Btw, here is the information from ZHA about this button (which works fine but is still of unknown type):

```json
{
  "node_descriptor": "NodeDescriptor()",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0000",
        "0x0001",
        "0x0003"
      ],
      "out_clusters": [
        "0x0000",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0008",
        "0x0019"
      ]
    }
  },
  "manufacturer": "LUMI",
  "model": "lumi.sensor_switch",
  "class": "zhaquirks.xiaomi.mija.sensor_switch.MijaButton"
}
```

while another identical button as a filled `NodeDescriptor`:


```json
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.EndDevice: 2>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.AllocateAddress: 128>, manufacturer_code=4151, maximum_buffer_size=127, maximum_incoming_transfer_size=100, server_mask=0, maximum_outgoing_transfer_size=100, descriptor_capability_field=<DescriptorCapability.0: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=True, *is_full_function_device=False, *is_mains_powered=False, *is_receiver_on_when_idle=False, *is_router=False, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": 260,
      "device_type": "0x0006",
      "in_clusters": [
        "0x0000",
        "0x0001",
        "0x0003"
      ],
      "out_clusters": [
        "0x0000",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0008",
        "0x0019"
      ]
    }
  },
  "manufacturer": "LUMI",
  "model": "lumi.sensor_switch",
  "class": "zhaquirks.xiaomi.mija.sensor_switch.MijaButton"
}
```

Reply:
> @puddly if there's a node_descriptor, can it have a logical type of None?

@samueltardieu's most recent comment is correct, this was fixed by zigpy/zigpy#773.

Do you want to make a new zigpy release incorporating that and the appdb schema rewrite (to fix the device deletion issues)?

Reply:
Yes. Also wanted to add that zcl 7 attribute for power restoration.
And some other quick fix for the probing function typing.